app/controllers/advanced_messenger_controller.rb

```diff
@@ -7,19 +7,26 @@ def update_read_by_users
       render_404
       return
     end
-    read_by_users = JSON.parse(@journal.read_by_users);
-    if read_by_users[User.current.id.to_s] == nil
-      Rails.logger.error("Change read status of journal #{params[:id]} for user #{User.current.id} is not allowed")
-      render_404
+    
+    if !@journal.journalized.visible? || @journal.private_notes? && !User.current.allowed_to?(:view_private_notes, @journal.journalized.project) && @journal.user.id =! User.current.id
+      Rails.logger.error("Current user doesn't have permissions to view journal #{params[:id]}")
+      render_403
       return
     end
-    if not ["0", "1", "2"].include? params[:read_by_users]
+    read_by_users = JSON.parse(@journal.read_by_users);
+    if not ["0", "1", "2", "3"].include? params[:read_by_users]
       Rails.logger.error("#{params[:read_by_users]} is not a valid journal read status for a user")
       render_404
       return
     end
 
-    read_by_users[User.current.id.to_s] = ReadByUsers.new params[:read_by_users].to_i
+    if read_by_users[User.current.id.to_s] == nil &&  params[:read_by_users] == "2"
+      read_by_users[User.current.id.to_s] = {collapsed: 1}
+    elsif read_by_users[User.current.id.to_s] != nil && read_by_users[User.current.id.to_s]["collapsed"] != nil && params[:read_by_users] == "1"
+      read_by_users.delete(User.current.id.to_s)
+    else  
+      read_by_users[User.current.id.to_s] = {read: params[:read_by_users].to_i, date: DateTime.now}
```

Comment:
So `0, 3`  means `read`?

Reply:
0 means read, 3 was a left over of a not implemneted "ignored" state. I removed it for the moment